File: src/visual-graph/builder/ensure-expression-statement-node.test.ts
import { describe, expect, test } from "vitest";

import type { Span } from "../../ir/primitive/span.js";
import { asReferenceId } from "../../ir/serialized/reference-id.js";
import type { SerializedHeadExpression } from "../../ir/serialized/serialized-expression-statement-head.js";
import type { SerializedReference } from "../../ir/serialized/serialized-reference.js";
import { NODE_KIND } from "../node-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualElement } from "../visual-element.js";
import type { VisualNode } from "../visual-node.js";
import type { BuildState } from "./build-state.js";
import { ensureExpressionStatementNode } from "./ensure-expression-statement-node.js";
import { baseRef } from "./testing/make-ref.js";
import { span } from "./testing/span.js";

function spanAt(offset: number, line: number): Span {
  return { offset, line, column: 0 };
}

function makeState(): BuildState {
  return {
    subgraphByScope: new Map(),
    functionSubgraphByFn: new Map(),
    returnSubgraphsByFn: new Map(),
    returnUseAdded: new Set(),
    ifTestAnchorByOffset: new Map(),
    switchDiscriminantAnchorByOffset: new Map(),
    whileTestAnchorByOffset: new Map(),
    doWhileTestAnchorByOffset: new Map(),
    forTestAnchorByOffset: new Map(),
    pendingLoopTestAnchors: [],
    expressionStatementByOffset: new Map(),
    emittedEdges: new Set(),
    edges: [],
  };
}

function refWithHead(opts: {
  refId: string;
  startOffset: number;
  endOffset: number;
  startLine: number;
  endLine?: number;
  head: SerializedHeadExpression;
}): SerializedReference {
  return {
    ...baseRef(),
    id: asReferenceId(opts.refId),
    identifier: { name: "x", span: span() },
    expressionStatementContainer: {
      startSpan: spanAt(opts.startOffset, opts.startLine),
      endSpan: spanAt(opts.endOffset, opts.endLine ?? opts.startLine),
      head: opts.head,
    },
  };
}

const CONSOLE_LOG_HEAD: SerializedHeadExpression = {
  kind: "call",
  callee: {
    kind: "member",
    object: { kind: "identifier", name: "console" },
    property: "log",
  },
};

describe("ensureExpressionStatementNode", () => {
  test("returns null when the ref carries no ExpressionStatement container", () => {
    const elements: VisualElement[] = [];
    const state = makeState();
    expect(
      ensureExpressionStatementNode(baseRef(), "", elements, state),
    ).toEqual(null);
    expect(elements).toHaveLength(0);
  });

  test("renders a call/member/identifier mini-AST to `<receiver>.<property>()`", () => {
    const ref = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 15,
      startLine: 7,
      head: CONSOLE_LOG_HEAD,
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    const id = ensureExpressionStatementNode(
      ref,
      "console.log(a);",
      elements,
      state,
    );
    expect(id).toEqual("expr_stmt_0");
    expect(elements).toHaveLength(1);
    const node = elements[0] as VisualNode;
    expect(node).toMatchObject({
      type: VISUAL_ELEMENT_TYPE.Node,
      id: "expr_stmt_0",
      kind: NODE_KIND.LegacyExpressionStatement,
      name: "console.log()",
      line: 7,
    });
  });

  test("renders a bare identifier head with no parens", () => {
    const ref = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 2,
      startLine: 2,
      head: { kind: "identifier", name: "a" },
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    ensureExpressionStatementNode(ref, "a;", elements, state);
    expect((elements[0] as VisualNode).name).toEqual("a");
  });

  test("renders an awaited chain to `await <chain>()`", () => {
    const ref = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 50,
      startLine: 1,
      endLine: 5,
      head: {
        kind: "await",
        argument: {
          kind: "call",
          callee: {
            kind: "member",
            object: {
              kind: "call",
              callee: {
                kind: "member",
                object: { kind: "identifier", name: "Promise" },
                property: "resolve",
              },
            },
            property: "then",
          },
        },
      },
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    ensureExpressionStatementNode(ref, "", elements, state);
    expect((elements[0] as VisualNode).name).toEqual(
      "await Promise.resolve().then()",
    );
  });

  test("slices the original source for a `raw` head", () => {
    const ref = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 6,
      startLine: 3,
      head: { kind: "raw", startSpan: spanAt(0, 3), endSpan: spanAt(5, 3) },
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    ensureExpressionStatementNode(ref, "x = 1;", elements, state);
    expect((elements[0] as VisualNode).name).toEqual("x = 1");
  });

  test("sets endLine when the statement spans multiple lines so the renderer shows L<start>-<end>", () => {
    const ref = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 20,
      startLine: 1,
      endLine: 3,
      head: CONSOLE_LOG_HEAD,
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    ensureExpressionStatementNode(
      ref,
      "console.log(\n  a,\n);",
      elements,
      state,
    );
    expect((elements[0] as VisualNode).line).toEqual(1);
    expect((elements[0] as VisualNode).endLine).toEqual(3);
  });

  test("returns the cached id and does not re-append when called twice for refs in the same statement", () => {
    const refA = refWithHead({
      refId: "r1",
      startOffset: 0,
      endOffset: 15,
      startLine: 7,
      head: CONSOLE_LOG_HEAD,
    });
    const refB = refWithHead({
      refId: "r2",
      startOffset: 0,
      endOffset: 15,
      startLine: 7,
      head: CONSOLE_LOG_HEAD,
    });
    const elements: VisualElement[] = [];
    const state = makeState();
    const raw = "console.log(a);";
    const idA = ensureExpressionStatementNode(refA, raw, elements, state);
    const idB = ensureExpressionStatementNode(refB, raw, elements, state);
    expect(idA).toEqual(idB);
    expect(elements).toHaveLength(1);
  });
});

File: crates/unsnarl-visual-graph/src/builder/ensure_expression_statement_node_test.rs
//! Sibling tests for [`ensure_expression_statement_node`].

use unsnarl_ir::primitive::SourceIndex;
use unsnarl_ir::serialized::SerializedExpressionStatementContainer;
use unsnarl_ir::serialized::SerializedHeadExpression;

use super::ensure_expression_statement_node;
use crate::builder::arena::{BuildArena, Container, ElementHandle};
use crate::builder::builder_fixtures::{base_serialized_reference, reference_id, span_offset_line};
use crate::builder::state::BuildState;
use crate::visual_element::VisualElement;
use crate::visual_node::{SyntheticNodeKind, VisualNode};
use unsnarl_ir::serialized::SerializedReference;

fn ref_with_head(
    ref_id_str: &str,
    start_offset: u32,
    end_offset: u32,
    start_line: u32,
    end_line: u32,
    head: SerializedHeadExpression,
) -> SerializedReference {
    let mut r = base_serialized_reference();
    r.id = reference_id(ref_id_str);
    r.identifier = unsnarl_ir::serialized::SerializedReferenceIdentifier::new(
        "x".to_string(),
        span_offset_line(0, 1),
    );
    r.expression_statement_container = Some(SerializedExpressionStatementContainer {
        start_span: span_offset_line(start_offset, start_line),
        end_span: span_offset_line(end_offset, end_line),
        head,
        expression_start_span: None,
    });
    r
}

fn console_log_head() -> SerializedHeadExpression {
    SerializedHeadExpression::Call {
        callee: Box::new(SerializedHeadExpression::Member {
            object: Box::new(SerializedHeadExpression::Identifier {
                name: "console".to_string(),
            }),
            property: "log".to_string(),
        }),
        start_span: span_offset_line(0, 1),
        end_span: span_offset_line(14, 1),
    }
}

fn finalize_one_node(arena: BuildArena) -> Option<VisualNode> {
    let elements = arena.finalize_root();
    if elements.is_empty() {
        return None;
    }
    match elements
        .into_iter()
        .next()
        .expect("test fixture produces at least one element")
    {
        VisualElement::Node(n) => Some(n),
        _ => None,
    }
}

#[test]
fn returns_none_when_ref_has_no_expression_statement_container() {
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    let r = base_serialized_reference();
    let id = ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build(""),
        Container::Root,
    );
    assert_eq!(id, None);
    assert!(arena.root_children.is_empty());
}

#[test]
fn renders_call_member_identifier_to_receiver_dot_property() {
    let r = ref_with_head("r1", 0, 15, 7, 7, console_log_head());
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    let id = ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build("console.log(a);"),
        Container::Root,
    );
    assert_eq!(id.as_deref(), Some("expr_stmt_0"));
    let node = finalize_one_node(arena).expect("one node");
    let VisualNode::Synthetic(s) = node else {
        panic!("expected synthetic");
    };
    assert_eq!(s.id, "expr_stmt_0");
    assert!(matches!(
        s.kind,
        SyntheticNodeKind::SyntheticExpressionStatement
    ));
    assert_eq!(s.name, "console.log()");
    assert_eq!(s.line, 7);
}

#[test]
fn renders_bare_identifier_head_without_parens() {
    let r = ref_with_head(
        "r1",
        0,
        2,
        2,
        2,
        SerializedHeadExpression::Identifier {
            name: "a".to_string(),
        },
    );
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build("a;"),
        Container::Root,
    );
    let VisualNode::Synthetic(s) = finalize_one_node(arena).expect("node") else {
        panic!("expected synthetic");
    };
    assert_eq!(s.name, "a");
}

#[test]
fn renders_awaited_chain() {
    let head = SerializedHeadExpression::Await {
        argument: Box::new(SerializedHeadExpression::Call {
            callee: Box::new(SerializedHeadExpression::Member {
                object: Box::new(SerializedHeadExpression::Call {
                    callee: Box::new(SerializedHeadExpression::Member {
                        object: Box::new(SerializedHeadExpression::Identifier {
                            name: "Promise".to_string(),
                        }),
                        property: "resolve".to_string(),
                    }),
                    start_span: span_offset_line(6, 1),
                    end_span: span_offset_line(23, 1),
                }),
                property: "then".to_string(),
            }),
            start_span: span_offset_line(6, 1),
            end_span: span_offset_line(31, 1),
        }),
    };
    let r = ref_with_head("r1", 0, 50, 1, 5, head);
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build(""),
        Container::Root,
    );
    let VisualNode::Synthetic(s) = finalize_one_node(arena).expect("node") else {
        panic!("expected synthetic");
    };
    assert_eq!(s.name, "await Promise.resolve().then()");
}

#[test]
fn slices_original_source_for_raw_head() {
    let head = SerializedHeadExpression::Raw {
        start_span: span_offset_line(0, 3),
        end_span: span_offset_line(5, 3),
    };
    let r = ref_with_head("r1", 0, 6, 3, 3, head);
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build("x = 1;"),
        Container::Root,
    );
    let VisualNode::Synthetic(s) = finalize_one_node(arena).expect("node") else {
        panic!("expected synthetic");
    };
    assert_eq!(s.name, "x = 1");
}

#[test]
fn sets_end_line_when_statement_spans_multiple_lines() {
    let r = ref_with_head("r1", 0, 20, 1, 3, console_log_head());
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    ensure_expression_statement_node(
        &mut arena,
        &mut state,
        &r,
        &SourceIndex::build("console.log(\n  a,\n);"),
        Container::Root,
    );
    let VisualNode::Synthetic(s) = finalize_one_node(arena).expect("node") else {
        panic!("expected synthetic");
    };
    assert_eq!(s.line, 1);
    assert_eq!(s.end_line, Some(3));
}

#[test]
fn returns_cached_id_and_does_not_re_append_for_refs_in_same_statement() {
    let ref_a = ref_with_head("r1", 0, 15, 7, 7, console_log_head());
    let ref_b = ref_with_head("r2", 0, 15, 7, 7, console_log_head());
    let mut arena = BuildArena::new();
    let mut state = BuildState::new();
    let index = SourceIndex::build("console.log(a);");
    let id_a =
        ensure_expression_statement_node(&mut arena, &mut state, &ref_a, &index, Container::Root);
    let id_b =
        ensure_expression_statement_node(&mut arena, &mut state, &ref_b, &index, Container::Root);
    assert_eq!(id_a, id_b);
    // Single node appended to root.
    assert_eq!(arena.root_children.len(), 1);
    assert!(matches!(arena.root_children[0], ElementHandle::Node(_)));
}
